### PR TITLE
Fix case where filesystem store future dropping causes issues

### PR DESCRIPTION
**Pull request by allada**
In a rare case where a file is uploaded, but while the rename() function is called the future is dropped the filesystem store can get into a bad state.

fixes: #495

- - -
This change is [Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/496)

### CODE REVIEW

**Review by allada — COMMENTED** [review]
+@adam-singer 

> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/496#-:-Nlomu0L6S-PCWE4F2gO:b-bn1n1c)* status: 0 of 3 files reviewed, all discussions resolved (waiting on @adam-singer)

File: nativelink-store/tests/filesystem_store_test.rs [hunk 1]
```diff
@@ -909,4 +910,92 @@ mod filesystem_store_tests {
 
         Ok(())
     }
+
+    /// Regression test for: https://github.com/TraceMachina/nativelink/issues/495.
+    #[tokio::test]
+    async fn update_file_future_drops_before_rename() -> Result<(), Error> {
```
Comment (MarcusSorealheis) [inline nativelink-store/tests/filesystem_store_test.rs:1025]: nice

**Review by MarcusSorealheis — APPROVED** [review]
LGTM

**Review by adam-singer — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/496)* status: all files reviewed, 3 unresolved discussions (waiting on @allada)

___
*[`nativelink-store/src/filesystem_store.rs` line 583 at r1](https://reviewable.io/reviews/TraceMachina/nativelink/496#-NlvZ7ms4aJQasBM8ENI:-NlvZ7ms4aJQasBM8ENJ:bgfr8bs) ([raw file](https://github.com/TraceMachina/nativelink/blob/bd61f02ca6f2383e1cc1527f382484aff6b5f5e9/nativelink-store/src/filesystem_store.rs#L583)):*
> ```XML
>                 let from_path = encoded_file_path.get_file_path();
>                 // Internally tokio spawns fs commands onto a blocking thread anyways.
>                 // Since we are already on a blocking thread, we just need the `fs` wrapper to manage
> ```

maybe spell out `std::fs::rename`.

Given its the std::fs::rename it manages open-file permit (via FnOnce) or is the `call_with_permit` missing here?
___
*[`nativelink-util/src/fs.rs` line 246 at r1](https://reviewable.io/reviews/TraceMachina/nativelink/496#-NlvYJ6a1mvBW8DPp9rc:-NlvYJ6a1mvBW8DPp9rd:b-dbus1m) ([raw file](https://github.com/TraceMachina/nativelink/blob/bd61f02ca6f2383e1cc1527f382484aff6b5f5e9/nativelink-util/src/fs.rs#L246)):*
> ```XML
> /// Acquire a permit from the open file semaphore and call a raw function.
> #[inline]
> pub async fn call_with_permit<R>(func: impl FnOnce() -> Result<R, Error>) -> Result<R, Error> {
> ```

Defined but not used?

**Review by adam-singer — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/496#-:-Nly2Swi51UxPk4eLNxm:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @allada)

**Review by adam-singer — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/496)* status: all files reviewed, 3 unresolved discussions (waiting on @allada)

**Review by adam-singer — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r2, all commit messages.
> Dismissed @MarcusSorealheis from a discussion.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/496#-:-NoOU7ha3G6F2FUZRyen:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 1 LGTMs obtained

**Review by adam-singer — APPROVED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/496)* status: 1 of 1 LGTMs obtained, and pending CI: Remote / large-ubuntu-22.04

**Review by adam-singer — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/496#-:-NoPDwiFCXXxBfXQtoMV:bnfp4nl)* status: 1 of 1 LGTMs obtained, and pending CI: Remote / large-ubuntu-22.04

**Pull request merged**